Added entries for list. (#2)
smlhelp/book: src/types/list.md

# List

> Lists are the SML type of ordered collections of objects. Notably, you can create lists of any kind of object, so `int list`, `string list`, and `bool list` are all valid types. `list` on its own is a type constructor (meaning that it makes new types out of old types), so it is not a valid type by itself, however. Lists are not, however, the same as arrays - they do not have constant-time access to any given index of the list. Instead, you only have access to the elements located at the very beginning of the list, the _head_. Additionally, lists are more _restricted_ than data structures in some other languages - a given list has a fixed type for its elements. All of the elements in a list must be of the same type. For the purposes of this document, we will discuss only int lists.

Reviewer: Is there an advantage to sticking to int lists? I think it might be a good idea to familiarize them with polymorphic structures early

Author: I asked Jacob, and he said that he'd recommend not introducing polymorphic lists yet. I was going for a page for lists that would be understandable to a student in the course right now, hence only discussion of int lists.

Reviewer: Well, we can discuss lists without needing to talk about `'a list` as a type. Instead of talking about polymorphism, we could just mention that int list, bool list, etc, are all types. I think it's important to show them early examples of multiple different list types (with bools, nested lists, etc).

Author: I've made a change that should have some mention of arbitrary typed lists now.

## Values

We write lists as a sequence of integers, separated by commas, all enclosed with two square brackets. So then we have valid int lists as `[1, 2, 3]`, `[1, 5, 1, 5, 0]`, and `[]`, with the latter representing the empty list. We also refer to the empty list as _nil_ (in addition, you can type `nil` instead of `[]` in code). 

> The other essential component to lists is what is known as the `::` operator, referred to as "cons".

Reviewer: Related to the above, I think it's a bit misleading to say that the type of cons is specialized to ints. It might be a better idea to describe its type as `t * t list -> t list` or something (for any type `t`).

Cons is also a _constructor_, meaning that it can be used to pattern match and deconstruct lists. As such, we can write basic functions to compute the length of a list as follows:
```sml
(* length : int list -> int *)
(* REQUIRES: true *)
(* ENSURES: length L ==> the length of L *)
fun length ([] : int list) : int = 0
  | length (x::xs : int list) : int = 1 + length xs
```
Given a non-empty list, this function simply binds the first element of the list to the identifier `x`, discards it, and then recursively calls `length` to find the length of the remaining list `xs`, adding 1 to the result.

The definition of an int list thus corresponds to:
```sml

> datatype int list = [] | :: of int * int list

Reviewer: Other than gripes with talking only about ints, this one is actually invalid syntax; while I'm all for lying to beginners, I think this is a bit of a line being crossed.

Author: That's why I wrote "corresponds to" - though I could add a disclaimer saying that this would not actually work in the SML/NJ REPL, afterwards. Thoughts?

```
where an int list can either be the empty list `[]`, or it can be `::` of a first element and the rest of the list (where `::` is an infix operator, so instead of being written as `::(x, xs)`, we have `x :: xs`).

## Motivation

Compared to other data structures, lists seem to have numerous disadvantages. As mentioned previously, they do not possess constant-time indexing like arrays in other languages - there is no way to instantly get the ith element of a list easily. Instead, you must "cons off" all the elements in front of that item in order to retrieve it - if you want to remove that item from the list, then you have to put the preceding elements _back_ as well (and in the right order!). Lists are also inherently sequential - you cannot access multiple elements at one time. 

> The reason why we choose lists is that they have very nice mathematical properties. These "disadvantages" in the previous paragraph become strengths, when viewed in a certain manner. Lists are powerful for their simple, inductive definition (as shown in the previous section), which is sufficiently powerful to characterize many important principles in this class. Our hope is that, throughout this course, you can begin to see that there is an elegance in simplicity.

Reviewer: These two paragraphs seem unconvincing to me. I'm not sure of a good way to immediately resolve that, but offhand I'd either avoid bringing it up at all, or mentioning some concrete ways that it isn't a disadvantage (persistence, desgining code flows such that it matters less, etc).

>
> > __Definition : Sorted Int Lists__
> > 
> >   1. `[] : int list` is sorted.
> >   2. The singleton int list is sorted.
> >   3. If `L : int list` is sorted, then if `x : int` is less than or equal to the first element of `L`, then `x :: L` is sorted.  
>
> This definition is naturally inductive, and follows very easily from the definition of sorting. In addition, it goes hand-in-hand with how we define lists - building them up from smaller parts one-by-one. Seen in this way, reasoning about and proving whether a list is sorted becomes very easy.

## Combination

We have seen that cons is essential for constructing lists, and for deconstructing the constituent elements that comprise a given list. What about when dealing with multiple lists? We might want to _combine_ the elements from several lists at once. A standard function for doing so is called the `@` operator, or "append".

```sml

> fun @ ([] : int list, R : int list) : int list = R

Reviewer: Again, invalid syntax.

# List

> Lists are the SML type of ordered collections of objects. Notably, you can create lists of any kind of object, so `int list`, `string list`, and `bool list` are all valid types. `list` on its own is a type constructor (meaning that it makes new types out of old types), so it is not a valid type by itself, however. Lists are not, however, the same as arrays - they do not have constant-time access to any given index of the list. Indeed, they are best thought of as not analogous to arrays in other languages at all. Instead, you only have access to the elements located at the very beginning of the list, the _head_. Additionally, lists are more _restricted_ than data structures in some other languages - a given list has a fixed type for its elements. All of the elements in a list must be of the same type. For the purposes of this document, we will discuss only int lists.

Reviewer 2: ```suggestion
Lists are the SML type of ordered collections of objects. Notably, you can create lists of any kind of object, so `int list`, `string list`, and `bool list` are all valid types. `list` on its own is a type constructor (meaning that it makes new types out of old types), so it is not a valid type by itself, however. Note that lists are not the same as arrays - they do not have constant-time access to any given index of the list. Indeed, they are best thought of as not analogous to arrays in other languages at all. Instead, you only have access to the elements located at the very beginning of the list, the _head_. Additionally, lists are more _restricted_ than data structures in some other languages - a given list has a fixed type for its elements. All of the elements in a list must be of the same type. For the purposes of this document, we will discuss only `int list`s.
```
Could it be worth splitting this paragraph right before talking about arrays? Also, I worry that the phrasing "you only have access to the elements located at the very beginning of the list" might be a bit confusing - you do have access to all the elements, but you have to go in order to get to them.

## Values

> We write lists as a sequence of integers, separated by commas, all enclosed with two square brackets. So then we have valid int lists as `[1, 2, 3]`, `[1, 5, 1, 5, 0]`, and `[]`, with the latter representing the empty list. We also refer to the empty list as _nil_ (in addition, you can type `nil` instead of `[]` in code). 

Reviewer 2: ```suggestion
We write lists as a sequence of integers, separated by commas, all enclosed with two square brackets. So then we have valid `int list`s as `[1, 2, 3]`, `[1, 5, 1, 5, 0]`, and `[]`, with the latter representing the empty list. We also refer to the empty list as _nil_ (in addition, you can type `nil` instead of `[]` in code). 
```

The other essential component to lists is what is known as the `::` operator, referred to as "cons". Cons can be used as a constructor for any fixed type of list, so we say that for any given type `t`:
```sml
(op ::) : t * t list -> t list
```

> Cons takes in a value `v` of type `t` and `t list`, and prepends `v` to the front of the given list. For ints in particular, we have that `1 :: [2, 3]` steps to `[1, 2, 3]`. Additionally, cons is _right-associative_. This means that in a continuous stream of applications of cons, they are evaluated from _right-to-left_. This means that `1 :: 2 :: 3 :: []` is implicitly denoting `1 :: (2 :: (3 :: []))`, as the calls to cons associate to the right. So, similarly to before, a `1 :: [2, 3]`, `[1, 2, 3]`, and `1 :: 2 :: 3 :: []` are exactly equivalent, and denote the same list.

Reviewer 2: ```suggestion
Cons takes in a value `v` of type `t` and `t list`, and prepends `v` to the front of the given list. For `int`s in particular, we have that `1 :: [2, 3]` steps to `[1, 2, 3]`. Additionally, cons is _right-associative_. This means that in a continuous stream of applications of cons, they are evaluated from _right-to-left_. This means that `1 :: 2 :: 3 :: []` is implicitly denoting `1 :: (2 :: (3 :: []))`, as the calls to cons associate to the right. So, similarly to before, a `1 :: [2, 3]`, `[1, 2, 3]`, and `1 :: 2 :: 3 :: []` are exactly equivalent, and denote the same list.
```

> Cons is also a _constructor_, meaning that it can be used to pattern match and deconstruct lists. As such, we can write basic functions to compute the length of a list as follows:

Reviewer 2: Should we note that nil is also a constructor?

```sml
(* length : int list -> int *)
(* REQUIRES: true *)
(* ENSURES: length L ==> the length of L *)
fun length ([] : int list) : int = 0
  | length (x::xs : int list) : int = 1 + length xs
```
Given a non-empty list, this function simply binds the first element of the list to the identifier `x`, discards it, and then recursively calls `length` to find the length of the remaining list `xs`, adding 1 to the result.

> The definition of an int list thus corresponds to:

Reviewer 2: ```suggestion
The definition of an `int list` thus corresponds to:
```

```sml
datatype int list = [] | :: of int * int list
```

> where an int list can either be the empty list `[]`, or it can be `::` of a first element and the rest of the list (where `::` is an infix operator, so instead of being written as `::(x, xs)`, we have `x :: xs`). Note that this is not actually valid syntax, but you can think of the definition of int lists in this way.

Reviewer 2: ```suggestion
where an `int list` can either be the empty list `[]`, or it can be `::` of a first element and the rest of the list (where `::` is an infix operator, so instead of being written as `:: (x,xs)`, we have `x :: xs`). Note that this is not actually valid syntax, but you can think of the definition of int lists in this way.
```
(Same change (int list to `int list`) in potentially other places also, but I noted the ones I saw)
